File: client/src/common/PositionComponentData.ts
import {EntityComponentData} from "../engine/simulation/EntityData";
import {Vector2Animation} from "../misc/Vector2Animation";
import {Entity} from "../engine/simulation/Entity";
import {EntityComponent} from "../engine/simulation/EntityComponent";
import {EntityComponentGetter} from "../engine/simulation/EntityComponentGetter";

export interface PositionComponentData extends EntityComponentData {
  positionAnimation: Vector2Animation
}

export const PositionComponent = new EntityComponentGetter<PositionComponentData>("PositionComponentData")

export function resolveEntityPositionForCurrentTime(entity: Entity) {
  const positionComponent = entity.getComponent<PositionComponentData>("PositionComponentData")
  return Vector2Animation.resolve(positionComponent.data.positionAnimation, entity.simulation.getCurrentSimulationTime())
}

export function resolveEntityPositionForTime(entity: Entity, simulationTime: number) {
  const positionComponent = entity.getComponent<PositionComponentData>("PositionComponentData")
  return Vector2Animation.resolve(positionComponent.data.positionAnimation, simulationTime)
}

export function resolvePositionForCurrentTime(positionComponent: EntityComponent<PositionComponentData>) {
  return Vector2Animation.resolve(positionComponent.data.positionAnimation, positionComponent.entity.simulation.getCurrentSimulationTime())
}

export function resolvePositionForTime(positionComponent: EntityComponent<PositionComponentData>, simulationTime: number) {
  return Vector2Animation.resolve(positionComponent.data.positionAnimation, simulationTime)
}

File: simulation-server/src/main/kotlin/botfarm/simulationserver/common/PositionComponentData.kt
package botfarm.simulationserver.common

import botfarm.misc.Vector2
import botfarm.misc.Vector2Animation
import botfarm.simulationserver.simulation.Entity
import botfarm.simulationserver.simulation.EntityComponentData

data class PositionComponentData(
   val positionAnimation: Vector2Animation = Vector2Animation()
) : EntityComponentData()

fun Entity.resolvePosition(): Vector2 {
   val positionComponent = this.getComponent<PositionComponentData>()
   return positionComponent.data.positionAnimation.resolve(this.simulation.getCurrentSimulationTime())
}

fun Entity.resolvePosition(simulationTime: Double): Vector2 {
   val positionComponent = this.getComponent<PositionComponentData>()
   return positionComponent.data.positionAnimation.resolve(simulationTime)
}